// mind-elixir/dist/MindElixirLite.js
const Rn = 0, Xn = 1, Vn = 2, et = {
  name: "Latte",
  type: "light",
  palette: ["#dd7878", "#ea76cb", "#8839ef", "#e64553", "#fe640b", "#df8e1d", "#40a02b", "#209fb5", "#1e66f5", "#7287fd"],
  cssVar: {
    "--node-gap-x": "30px",
    "--node-gap-y": "10px",
    "--main-gap-x": "65px",
    "--main-gap-y": "45px",
    "--root-radius": "30px",
    "--main-radius": "20px",
    "--root-color": "#ffffff",
    "--root-bgcolor": "#4c4f69",
    "--root-border-color": "rgba(0, 0, 0, 0)",
    "--main-border": "",
    // you can customize, it will fallback to 2px solid main-color
    "--main-color": "#444446",
    "--main-bgcolor": "#ffffff",
    "--main-bgcolor-transparent": "rgba(255, 255, 255, 0.8)",
    "--topic-padding": "3px",
    "--color": "#777777",
    "--bgcolor": "#f6f6f6",
    "--selected": "#4dc4ff",
    "--accent-color": "#e64553",
    "--panel-color": "#444446",
    "--panel-bgcolor": "#ffffff",
    "--panel-border-color": "#eaeaea",
    "--map-padding": "50px 80px"
  }
}, nt = {
  name: "Dark",
  type: "dark",
  palette: ["#848FA0", "#748BE9", "#D2F9FE", "#4145A5", "#789AFA", "#706CF4", "#EF987F", "#775DD5", "#FCEECF", "#DA7FBC"],
  cssVar: {
    "--node-gap-x": "30px",
    "--node-gap-y": "10px",
    "--main-gap-x": "65px",
    "--main-gap-y": "45px",
    "--root-radius": "30px",
    "--main-radius": "20px",
    "--root-color": "#ffffff",
    "--root-bgcolor": "#2d3748",
    "--root-border-color": "rgba(255, 255, 255, 0.1)",
    "--main-border": "",
    "--main-color": "#ffffff",
    "--main-bgcolor": "#4c4f69",
    "--main-bgcolor-transparent": "rgba(76, 79, 105, 0.8)",
    "--topic-padding": "3px",
    "--color": "#cccccc",
    "--bgcolor": "#252526",
    "--selected": "#4dc4ff",
    "--accent-color": "#789AFA",
    "--panel-color": "#ffffff",
    "--panel-bgcolor": "#2d3748",
    "--panel-border-color": "#696969",
    "--map-padding": "50px 80px"
  }
};
function Ot(t) {
  return t.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/"/g, "&quot;");
}
const j = function(t, e) {
  if (e.id === t)
    return e;
  if (e.children && e.children.length) {
    for (let n = 0; n < e.children.length; n++) {
      const o = j(t, e.children[n]);
      if (o) return o;
    }
    return null;
  } else
    return null;
}, ot = (t, e) => {
  if (t.parent = e, t.children)
    for (let n = 0; n < t.children.length; n++)
      ot(t.children[n], t);
}, bt = (t, e, n) => {
  t.expanded = e, t.children && t.children.forEach((o) => {
    bt(o, e);
  });
};
function _(t, e, n, o) {
  const s = n - t, i = o - e, c = Math.atan2(i, s) * 180 / Math.PI, a = 12, d = 30, f = (c + 180 - d) * Math.PI / 180, l = (c + 180 + d) * Math.PI / 180;
  return {
    x1: n + Math.cos(f) * a,
    y1: o + Math.sin(f) * a,
    x2: n + Math.cos(l) * a,
    y2: o + Math.sin(l) * a
  };
}
function X() {
  return ((/* @__PURE__ */ new Date()).getTime().toString(16) + Math.random().toString(16).substring(2)).substring(2, 18);
}
const Bt = function() {
  const t = X();
  return {
    topic: this.newTopicName,
    id: t
  };
};
function Yt(t) {
  return JSON.parse(
    JSON.stringify(t, (n, o) => {
      if (n !== "parent")
        return o;
    })
  );
}
const P = (t, e) => {
  let n = 0, o = 0;
  for (; e && e !== t; )
    n += e.offsetLeft, o += e.offsetTop, e = e.offsetParent;
  return { offsetLeft: n, offsetTop: o };
}, D = (t, e) => {
  for (const n in e)
    t.setAttribute(n, e[n]);
}, J = (t) => t ? t.tagName === "ME-TPC" : !1, st = (t) => {
  const e = /translate3d\(([^,]+),\s*([^,]+)/, n = t.match(e);
  return n ? { x: parseFloat(n[1]), y: parseFloat(n[2]) } : { x: 0, y: 0 };
}, vt = function(t) {
  for (let e = 0; e < t.length; e++) {
    const { dom: n, evt: o, func: s } = t[e];
    n.addEventListener(o, s);
  }
  return function() {
    for (let n = 0; n < t.length; n++) {
      const { dom: o, evt: s, func: i } = t[n];
      o.removeEventListener(s, i);
    }
  };
}, dt = (t, e) => {
  const n = t.x - e.x, o = t.y - e.y;
  return Math.sqrt(n * n + o * o);
}, Wt = function(t, e) {
  if (!e)
    return q(t), t;
  let n = t.querySelector(".insert-preview");
  const o = `insert-preview ${e} show`;
  return n || (n = document.createElement("div"), t.appendChild(n)), n.className = o, t;
}, q = function(t) {
  if (!t) return;
  const e = t.querySelectorAll(".insert-preview");
  for (const n of e || [])
    n.remove();
}, ft = function(t, e) {
  for (const n of e) {
    const o = n.parentElement.parentElement.contains(t);
    if (!(t && t.tagName === "ME-TPC" && t !== n && !o && t.nodeObj.parent)) return !1;
  }
  return !0;
}, Rt = function(t) {
  const e = document.createElement("div");
  return e.className = "mind-elixir-ghost", t.container.appendChild(e), e;
};
class Xt {
  mind;
  isMoving = !1;
  interval = null;
  speed = 20;
  constructor(e) {
    this.mind = e;
  }
  move(e, n) {
    this.isMoving || (this.isMoving = !0, this.interval = setInterval(() => {
      this.mind.move(e * this.speed * this.mind.scaleVal, n * this.speed * this.mind.scaleVal);
    }, 100));
  }
  stop() {
    this.isMoving = !1, this.interval && (clearInterval(this.interval), this.interval = null);
  }
}
function Vt(t) {
  return {
    isDragging: !1,
    insertType: null,
    meet: null,
    ghost: Rt(t),
    edgeMoveController: new Xt(t),
    startX: 0,
    startY: 0,
    pointerId: null
  };
}
const zt = 5;
function ut(t, e, n, o = !1) {
  if (t.spacePressed) return !1;
  const s = n.target;
  if (s?.tagName !== "ME-TPC" || !s.nodeObj.parent) return !1;
  if (e.startX = n.clientX, e.startY = n.clientY, e.pointerId = n.pointerId, t.dragged = t.currentNodes, o) {
    wt(t, e);
    const i = t.container.getBoundingClientRect();
    xt(e.ghost, n.clientX - i.x, n.clientY - i.y);
  }
  return !0;
}
function xt(t, e, n) {
  t.style.transform = `translate(${e - 10}px, ${n - 10}px)`, t.style.display = "block";
}
function wt(t, e) {
  const { dragged: n } = t;
  if (!n) return;
  const o = document.activeElement;
  o && o.isContentEditable && o.blur(), e.isDragging = !0, n.length > 1 ? e.ghost.innerHTML = n.length + "" : e.ghost.innerHTML = n[0].innerHTML;
  for (const s of n)
    s.parentElement.parentElement.style.opacity = "0.5";
  t.panHelper.clear();
}
function Ft(t, e, n) {
  const { dragged: o } = t;
  if (!o || e.pointerId !== n.pointerId) return;
  const s = n.clientX - e.startX, i = n.clientY - e.startY, r = Math.sqrt(s * s + i * i);
  if (!e.isDragging && r > zt && wt(t, e), !e.isDragging) return;
  const c = t.container.getBoundingClientRect();
  xt(e.ghost, n.clientX - c.x, n.clientY - c.y), n.clientX < c.x + 50 ? e.edgeMoveController.move(1, 0) : n.clientX > c.x + c.width - 50 ? e.edgeMoveController.move(-1, 0) : n.clientY < c.y + 50 ? e.edgeMoveController.move(0, 1) : n.clientY > c.y + c.height - 50 ? e.edgeMoveController.move(0, -1) : e.edgeMoveController.stop(), q(e.meet);
  const a = 12 * t.scaleVal, d = document.elementFromPoint(n.clientX, n.clientY - a);
  if (ft(d, o)) {
    e.meet = d;
    const f = d.getBoundingClientRect(), l = f.y;
    n.clientY > l + f.height ? e.insertType = "after" : e.insertType = "in";
  } else {
    const f = document.elementFromPoint(n.clientX, n.clientY + a);
    if (ft(f, o)) {
      e.meet = f;
      const u = f.getBoundingClientRect().y;
      n.clientY < u ? e.insertType = "before" : e.insertType = "in";
    } else
      e.insertType = null, e.meet = null;
  }
  e.meet && Wt(e.meet, e.insertType);
}
function Gt(t, e, n) {
  const { dragged: o } = t;
  if (!(!o || e.pointerId !== n.pointerId)) {
    e.edgeMoveController.stop();
    for (const s of o)
      s.parentElement.parentElement.style.opacity = "1";
    e.ghost.style.display = "none", e.ghost.innerHTML = "", e.isDragging && e.meet && (q(e.meet), e.insertType === "before" ? t.moveNodeBefore(o, e.meet) : e.insertType === "after" ? t.moveNodeAfter(o, e.meet) : e.insertType === "in" && t.moveNodeIn(o, e.meet)), t.dragged = null, e.isDragging = !1, e.insertType = null, e.meet = null, e.pointerId = null;
  }
}
function pt(t, e) {
  const { dragged: n } = t;
  if (n) {
    e.edgeMoveController.stop();
    for (const o of n)
      o.parentElement.parentElement.style.opacity = "1";
    e.meet && q(e.meet), e.ghost.style.display = "none", e.ghost.innerHTML = "", t.dragged = null, e.isDragging = !1, e.insertType = null, e.meet = null, e.pointerId = null;
  }
}
const B = {
  LHS: "lhs",
  RHS: "rhs"
}, jt = function() {
  this.nodes.innerHTML = "";
  const t = this.createTopic(this.nodeData);
  St.call(this, t, this.nodeData), t.draggable = !1;
  const e = document.createElement("me-root");
  e.appendChild(t);
  const n = this.nodeData.children || [];
  if (this.direction === 2) {
    let o = 0, s = 0;
    n.map((i) => {
      i.direction === 0 ? o += 1 : i.direction === 1 ? s += 1 : o <= s ? (i.direction = 0, o += 1) : (i.direction = 1, s += 1);
    });
  }
  _t(this, n, e);
}, _t = function(t, e, n) {
  const o = document.createElement("me-main");
  o.className = B.LHS;
  const s = document.createElement("me-main");
  s.className = B.RHS;
  for (let i = 0; i < e.length; i++) {
    const r = e[i], { grp: c } = t.createWrapper(r);
    t.direction === 2 ? r.direction === 0 ? o.appendChild(c) : s.appendChild(c) : t.direction === 0 ? o.appendChild(c) : s.appendChild(c);
  }
  t.nodes.appendChild(o), t.nodes.appendChild(n), t.nodes.appendChild(s), t.nodes.appendChild(t.lines), t.nodes.appendChild(t.labelContainer);
}, qt = function(t, e) {
  const n = document.createElement("me-children");
  for (let o = 0; o < e.length; o++) {
    const s = e[o], { grp: i } = t.createWrapper(s);
    n.appendChild(i);
  }
  return n;
}, Ct = function(t, e) {
  const o = (this?.el ? this.el : e || document).querySelector(`[data-nodeid="me${t}"]`);
  if (!o) throw new Error(`FindEle: Node ${t} not found, maybe it's collapsed.`);
  return o;
}, St = function(t, e) {
  if (t.innerHTML = "", e.style) {
    const n = e.style;
    for (const o in n)
      t.style[o] = n[o];
  }
  if (e.dangerouslySetInnerHTML) {
    t.innerHTML = e.dangerouslySetInnerHTML;
    return;
  }
  if (e.image) {
    const n = e.image;
    if (n.url && n.width && n.height) {
      const o = document.createElement("img");
      o.src = this.imageProxy ? this.imageProxy(n.url) : n.url, o.style.width = n.width + "px", o.style.height = n.height + "px", n.fit && (o.style.objectFit = n.fit), t.appendChild(o), t.image = o;
    }
  } else t.image && (t.image = void 0);
  {
    const n = document.createElement("span");
    n.className = "text", this.markdown ? n.innerHTML = this.markdown(e.topic, e) : n.textContent = e.topic, t.appendChild(n), t.text = n;
  }
  if (e.hyperLink) {
    const n = document.createElement("a");
    n.className = "hyper-link", n.target = "_blank", n.innerText = "🔗", n.href = e.hyperLink, t.appendChild(n), t.link = n;
  } else t.link && (t.link = void 0);
  if (e.icons && e.icons.length) {
    const n = document.createElement("span");
    n.className = "icons", n.innerHTML = e.icons.map((o) => `<span>${Ot(o)}</span>`).join(""), t.appendChild(n), t.icons = n;
  } else t.icons && (t.icons = void 0);
  if (e.tags && e.tags.length) {
    const n = document.createElement("div");
    n.className = "tags", e.tags.forEach((o) => {
      const s = document.createElement("span");
      typeof o == "string" ? s.textContent = o : (s.textContent = o.text, o.className && (s.className = o.className), o.style && Object.assign(s.style, o.style)), n.appendChild(s);
    }), t.appendChild(n), t.tags = n;
  } else t.tags && (t.tags = void 0);
}, Ut = function(t, e) {
  const n = document.createElement("me-wrapper"), { p: o, tpc: s } = this.createParent(t);
  if (n.appendChild(o), !e && t.children && t.children.length > 0) {
    const i = te(t.expanded);
    if (o.appendChild(i), t.expanded !== !1) {
      const r = qt(this, t.children);
      n.appendChild(r);
    }
  }
  return { grp: n, top: o, tpc: s };
}, Kt = function(t) {
  const e = document.createElement("me-parent"), n = this.createTopic(t);
  return St.call(this, n, t), e.appendChild(n), { p: e, tpc: n };
}, Jt = function(t) {
  const e = document.createElement("me-children");
  return e.append(...t), e;
}, Zt = function(t) {
  const e = document.createElement("me-tpc");
  return e.nodeObj = t, e.dataset.nodeid = "me" + t.id, e;
};
function Et(t) {
  const e = document.createRange();
  e.selectNodeContents(t);
  const n = window.getSelection();
  n && (n.removeAllRanges(), n.addRange(e));
}
const Qt = function(t) {
  if (!t) return;
  const e = document.createElement("div"), n = t.nodeObj, o = n.topic, { offsetLeft: s, offsetTop: i } = P(this.nodes, t);
  this.nodes.appendChild(e), e.id = "input-box", e.textContent = o, e.contentEditable = "plaintext-only", e.spellcheck = !1;
  const r = getComputedStyle(t);
  e.style.cssText = `
  left: ${s}px;
  top: ${i}px;
  min-width:${t.offsetWidth - 8}px;
  color:${r.color};
  font-size:${r.fontSize};
  padding:${r.padding};
  margin:${r.margin}; 
  background-color:${r.backgroundColor !== "rgba(0, 0, 0, 0)" && r.backgroundColor};
  border: ${r.border};
  border-radius:${r.borderRadius}; `, this.direction === 0 && (e.style.right = "0"), t.style.opacity = "0", Et(e), this.bus.fire("operation", {
    name: "beginEdit",
    obj: t.nodeObj
  }), e.addEventListener("keydown", (c) => {
    if (c.stopPropagation(), c.isComposing) return;
    const a = c.key;
    if (a === "Enter" || a === "Tab") {
      if (c.shiftKey) return;
      c.preventDefault(), e.blur(), this.container.focus();
    } else a === "Escape" && (c.preventDefault(), e.textContent = o, e.blur(), this.container.focus());
  }), e.addEventListener("blur", () => {
    if (!e) return;
    t.style.opacity = "1", e.remove();
    const c = e.innerText?.trim() || "";
    c === o || c === "" || (n.topic = c, this.markdown ? t.text.innerHTML = this.markdown(n.topic, n) : t.text.textContent = c, this.linkDiv(), this.bus.fire("operation", {
      name: "finishEdit",
      obj: n,
      origin: o
    }));
  });
}, te = function(t) {
  const e = document.createElement("me-epd");
  return e.expanded = t !== !1, e.className = t !== !1 ? "minus" : "", e;
}, ee = function(t) {
  const n = t.parentElement.parentElement.lastElementChild;
  n?.tagName === "svg" && n?.remove();
};
function ne(t) {
  return {
    nodeData: t.isFocusMode ? t.nodeDataBackup : t.nodeData,
    arrows: t.arrows,
    summaries: t.summaries,
    direction: t.direction,
    theme: t.theme,
    compact: t.compact,
    meta: t.meta
  };
}
const oe = function(t, e = !1) {
  const n = this.container, o = t.getBoundingClientRect(), s = n.getBoundingClientRect();
  if (e || o.top > s.bottom - 50 || o.bottom < s.top + 50 || o.left > s.right - 50 || o.right < s.left + 50) {
    const r = o.left + o.width / 2, c = o.top + o.height / 2, a = s.left + s.width / 2, d = s.top + s.height / 2, f = r - a, l = c - d;
    this.move(-f, -l, !0);
  }
}, se = function(t, e, n) {
  this.clearSelection(), this.scrollIntoView(t), this.selection?.select(t), e && this.bus.fire("selectNewNode", t.nodeObj);
}, ie = function(t) {
  this.selection?.select(t);
}, re = function(t) {
  this.selection?.deselect(t);
}, ce = function() {
  this.unselectNodes(this.currentNodes), this.unselectSummary(), this.unselectArrow();
}, Tt = function(t) {
  return JSON.stringify(t, (e, n) => {
    if (!(e === "parent" && typeof n != "string"))
      return n;
  });
}, le = function() {
  const t = ne(this);
  return Tt(t);
}, ae = function() {
  return JSON.parse(this.getDataString());
}, he = function() {
  this.editable = !0;
}, de = function() {
  this.editable = !1;
}, fe = function(t, e = { x: 0, y: 0 }) {
  if (t < this.scaleMin && t < this.scaleVal || t > this.scaleMax && t > this.scaleVal) return;
  const n = this.container.getBoundingClientRect(), o = e.x ? e.x - n.left - n.width / 2 : 0, s = e.y ? e.y - n.top - n.height / 2 : 0, { dx: i, dy: r } = it(this), c = this.map.style.transform, { x: a, y: d } = st(c), f = a - i, l = d - r, u = this.scaleVal, g = (-o + f) * (1 - t / u), y = (-s + l) * (1 - t / u);
  this.map.style.transform = `translate3d(${a - g}px, ${d - y}px, 0) scale(${t})`, this.scaleVal = t, this.bus.fire("scale", t);
}, ue = function() {
  const t = this.nodes.offsetHeight / this.container.offsetHeight, e = this.nodes.offsetWidth / this.container.offsetWidth, n = 1 / Math.max(1, Math.max(t, e));
  this.scaleVal = n;
  const { dx: o, dy: s } = it(this, !0);
  this.map.style.transform = `translate3d(${o}px, ${s}px, 0) scale(${n})`, this.bus.fire("scale", n);
}, pe = function(t, e, n = !1) {
  const { map: o, scaleVal: s, bus: i, container: r, nodes: c } = this;
  if (n && o.style.transition === "transform 0.3s")
    return;
  const a = o.style.transform;
  let { x: d, y: f } = st(a);
  const l = r.getBoundingClientRect(), u = c.getBoundingClientRect(), g = u.left < l.right && u.right > l.left, y = u.top < l.bottom && u.bottom > l.top;
  if (g) {
    const p = u.left + t, b = u.right + t;
    (p >= l.right || b <= l.left) && (t = 0);
  }
  if (y) {
    const p = u.top + e, b = u.bottom + e;
    (p >= l.bottom || b <= l.top) && (e = 0);
  }
  d += t, f += e, n && (o.style.transition = "transform 0.3s", setTimeout(() => {
    o.style.transition = "none";
  }, 300)), o.style.transform = `translate3d(${d}px, ${f}px, 0) scale(${s})`, i.fire("move", { dx: t, dy: e });
}, it = (t, e = !1) => {
  const { container: n, map: o, nodes: s } = t;
  let i, r;
  if (t.alignment === "nodes" || e)
    i = (n.offsetWidth - s.offsetWidth) / 2, r = (n.offsetHeight - s.offsetHeight) / 2, o.style.transformOrigin = "50% 50%";
  else {
    const c = o.querySelector("me-root"), a = c.offsetTop, d = c.offsetLeft, f = c.offsetWidth, l = c.offsetHeight;
    i = n.offsetWidth / 2 - d - f / 2, r = n.offsetHeight / 2 - a - l / 2, o.style.transformOrigin = `${d + f / 2}px 50%`;
  }
  return { dx: i, dy: r };
}, ge = function() {
  const { map: t, container: e } = this, { dx: n, dy: o } = it(this);
  e.scrollTop = 0, e.scrollLeft = 0, t.style.transform = `translate3d(${n}px, ${o}px, 0) scale(${this.scaleVal})`;
}, me = function(t) {
  t(this);
}, ye = function(t) {
  t.nodeObj.parent && (this.clearSelection(), this.tempDirection === null && (this.tempDirection = this.direction), this.isFocusMode || (this.nodeDataBackup = this.nodeData, this.isFocusMode = !0), this.nodeData = t.nodeObj, this.initRight(), this.toCenter());
}, be = function() {
  this.isFocusMode = !1, this.tempDirection !== null && (this.nodeData = this.nodeDataBackup, this.direction = this.tempDirection, this.tempDirection = null, this.refresh(), this.toCenter());
}, ve = function() {
  this.direction = 0, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, xe = function() {
  this.direction = 1, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, we = function() {
  this.direction = 2, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, Ce = function(t, e) {
  const n = t.nodeObj;
  typeof e == "boolean" ? n.expanded = e : n.expanded !== !1 ? n.expanded = !1 : n.expanded = !0;
  const o = t.getBoundingClientRect(), s = {
    x: o.left,
    y: o.top
  }, i = t.parentNode, r = i.children[1];
  if (r.expanded = n.expanded, r.className = n.expanded ? "minus" : "", ee(t), n.expanded) {
    const l = this.createChildren(
      n.children.map((u) => this.createWrapper(u).grp)
    );
    i.parentNode.appendChild(l);
  } else
    i.parentNode.children[1].remove();
  this.linkDiv(t.closest("me-main > me-wrapper"));
  const c = t.getBoundingClientRect(), a = {
    x: c.left,
    y: c.top
  }, d = s.x - a.x, f = s.y - a.y;
  this.move(d, f), this.bus.fire("expandNode", n);
}, Se = function(t, e) {
  const n = t.nodeObj, o = t.getBoundingClientRect(), s = {
    x: o.left,
    y: o.top
  };
  bt(n, e ?? !n.expanded), this.refresh();
  const i = this.findEle(n.id).getBoundingClientRect(), r = {
    x: i.left,
    y: i.top
  }, c = s.x - r.x, a = s.y - r.y;
  this.move(c, a);
}, Ee = function(t) {
  this.clearSelection(), t && (t = JSON.parse(JSON.stringify(t)), this.nodeData = t.nodeData, this.arrows = t.arrows || [], this.summaries = t.summaries || [], t.meta && (this.meta = t.meta), t.theme && this.changeTheme(t.theme)), ot(this.nodeData), this.layout(), this.linkDiv();
}, Te = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  cancelFocus: be,
  clearSelection: ce,
  disableEdit: de,
  enableEdit: he,
  expandNode: Ce,
  expandNodeAll: Se,
  focusNode: ye,
  getData: ae,
  getDataString: le,
  initLeft: ve,
  initRight: xe,
  initSide: we,
  install: me,
  move: pe,
  refresh: Ee,
  scale: fe,
  scaleFit: ue,
  scrollIntoView: oe,
  selectNode: se,
  selectNodes: ie,
  stringifyData: Tt,
  toCenter: ge,
  unselectNodes: re
}, Symbol.toStringTag, { value: "Module" })), De = 40, Le = 10, Me = ({ deltaMode: t, deltaY: e, viewportHeight: n }) => t === WheelEvent.DOM_DELTA_LINE ? e * De : t === WheelEvent.DOM_DELTA_PAGE ? e * n : e, Pe = ({ deltaMode: t, deltaY: e, scaleSensitivity: n, viewportHeight: o }) => {
  const i = -Me({ deltaMode: t, deltaY: e, viewportHeight: o }) / Le * n;
  return Math.max(-n, Math.min(n, i));
}, ke = (t, e, n) => {
  e !== 0 && t.scale(t.scaleVal + e, n);
}, Ae = (t, e) => {
  const n = Pe({
    deltaMode: e.deltaMode,
    deltaY: e.deltaY,
    scaleSensitivity: t.scaleSensitivity,
    viewportHeight: t.container.clientHeight || window.innerHeight
  });
  ke(t, n, { x: e.clientX, y: e.clientY });
};
function Ne(t) {
  const { panHelper: e, container: n } = t;
  let o = null;
  t.spacePressed = !1;
  const s = {
    lastTap: 0,
    lastTapTarget: null,
    DOUBLE_CLICK_THRESHOLD: 300,
    detect(h, m) {
      if (h.button !== 0) {
        this.clear();
        return;
      }
      const w = (/* @__PURE__ */ new Date()).getTime(), E = w - this.lastTap, S = E < this.DOUBLE_CLICK_THRESHOLD && E > 0 && this.lastTapTarget === h.target;
      this.lastTap = w, this.lastTapTarget = h.target, S && m(h);
    },
    clear() {
      this.lastTap = 0, this.lastTapTarget = null;
    }
  }, i = {
    Idle: 0,
    Pinch: 1,
    DragWait: 2,
    Drag: 3,
    Pan: 4,
    BoxSelect: 5
  };
  t.ptState = i.Idle;
  const r = {
    lastDistance: null,
    activePointers: /* @__PURE__ */ new Map(),
    handlePointerDown(h) {
      if (h.pointerType !== "touch") return !1;
      if (this.activePointers.set(h.pointerId, { x: h.clientX, y: h.clientY }), this.activePointers.size >= 2) {
        const [m, w] = Array.from(this.activePointers.values());
        return this.lastDistance = dt(m, w), !0;
      }
      return !1;
    },
    handlePointerMove(h) {
      if (h.pointerType !== "touch" || !this.activePointers.has(h.pointerId)) return !1;
      if (this.activePointers.set(h.pointerId, { x: h.clientX, y: h.clientY }), this.activePointers.size >= 2) {
        const [m, w] = Array.from(this.activePointers.values()), E = dt(m, w);
        if (this.lastDistance !== null && this.lastDistance > 0) {
          const S = E / this.lastDistance;
          t.scale(t.scaleVal * S, {
            x: (m.x + w.x) / 2,
            y: (m.y + w.y) / 2
          });
        }
        return this.lastDistance = E, !0;
      }
      return !1;
    },
    handlePointerUp(h) {
      h.pointerType === "touch" && (this.activePointers.delete(h.pointerId), this.activePointers.size < 2 && (this.lastDistance = null));
    },
    clear() {
      this.activePointers.clear(), this.lastDistance = null;
    }
  }, c = Vt(t), a = {
    timer: null,
    startPos: null,
    pointerId: null,
    DURATION: 500,
    MOVE_THRESHOLD: 10,
    clear() {
      this.timer !== null && (clearTimeout(this.timer), this.timer = null, this.startPos = null, this.pointerId = null);
    },
    start(h, m) {
      this.timer = window.setTimeout(() => {
        m(h), this.timer = null, this.startPos = null, this.pointerId = null;
      }, this.DURATION), this.startPos = { x: h.clientX, y: h.clientY }, this.pointerId = h.pointerId;
    },
    handleMove(h) {
      if (this.timer !== null && this.startPos !== null && h.pointerId === this.pointerId) {
        const m = h.clientX - this.startPos.x, w = h.clientY - this.startPos.y;
        Math.sqrt(m * m + w * w) > this.MOVE_THRESHOLD && this.clear();
      }
    }
  }, d = (h, m) => {
    if (h.closest("#input-box")) return !1;
    const w = h.closest(".svg-label"), E = h.closest(".topiclinks, .summary"), S = w ? { type: w.dataset.type, element: document.getElementById(w.dataset.svgId) } : E ? { type: E.classList.contains("topiclinks") ? "arrow" : "summary", element: h.closest("g") } : null;
    if (!S?.type || !S?.element) return !1;
    const { type: M, element: A } = S;
    return t.clearSelection(), M === "arrow" ? m ? t.editArrowLabel(A) : t.selectArrow(A) : m ? t.editSummary(A) : t.selectSummary(A), !0;
  }, f = (h) => {
    if (h.pointerType === "mouse" && h.button !== 0) return;
    if (t.helper1?.moved) {
      t.helper1.clear();
      return;
    }
    if (t.helper2?.moved) {
      t.helper2.clear();
      return;
    }
    if (e.moved) {
      e.clear();
      return;
    }
    if (c?.isDragging)
      return;
    const m = h.target;
    m.tagName === "ME-EPD" && (h.ctrlKey || h.metaKey ? t.expandNodeAll(m.previousSibling) : t.expandNode(m.previousSibling));
  }, l = (h) => {
    if (!t.editable) return;
    const m = h.target;
    if (J(m)) {
      t.selectNode(m), t.beginEdit(m);
      return;
    }
    d(m, !0);
  }, u = (h) => {
    if (h.pointerType === "touch" && r.handlePointerDown(h)) {
      t.ptState = i.Pinch, a.clear(), e.clear(), (c.isDragging || c.pointerId !== null) && pt(t, c);
      return;
    }
    if (t.ptState === i.Pinch) return;
    const m = h.target;
    if (t.editable && m.className === "map-container" && h.button === 0 && h.pointerType === "mouse") {
      t.ptState = i.BoxSelect;
      return;
    }
    if (e.handlePointerDown(h), e.mousedown && (t.ptState = i.Pan), h.button === 0 || h.pointerType === "touch")
      if (J(m)) {
        t.selection?.cancel();
        const E = t.currentNodes || [];
        if (h.ctrlKey || h.metaKey || t.mobileMultiSelect ? E.includes(m) ? o = m : ((t.currentArrow || t.currentSummary) && t.clearSelection(), t.selection?.select(m)) : E.includes(m) || t.selectNode(m), !t.editable) return;
        h.pointerType === "touch" ? (t.ptState = i.DragWait, a.start(h, (M) => {
          ut(t, c, M, !0) && (t.ptState = i.Drag, m.setPointerCapture(M.pointerId));
        })) : ut(t, c, h, !1) && (t.ptState = i.Drag, m.setPointerCapture(h.pointerId));
      } else
        d(m, !1);
  }, g = (h) => {
    switch (t.ptState) {
      case i.Pinch:
        r.handlePointerMove(h);
        break;
      case i.DragWait:
        a.handleMove(h), a.timer === null && (t.ptState = i.Pan, e.handlePointerMove(h));
        break;
      case i.Drag:
        Ft(t, c, h);
        break;
      case i.Pan:
        e.handlePointerMove(h);
        break;
    }
  }, y = (h) => {
    h.pointerType === "touch" && r.handlePointerUp(h);
    const m = c.isDragging, w = e.moved;
    switch (t.ptState) {
      case i.DragWait:
        a.clear();
        break;
      case i.Drag:
        Gt(t, c, h);
        break;
      case i.Pan:
        e.handlePointerUp(h);
        break;
    }
    s.detect(h, l), (t.ptState !== i.Pinch || r.activePointers.size < 2) && (t.ptState = i.Idle), o && (!m && !w && t.selection?.deselect(o), o = null);
  }, p = () => {
    r.clear(), a.clear(), e.clear(), s.clear(), (c.isDragging || c.pointerId !== null) && pt(t, c), t.ptState = i.Idle, o = null;
  }, b = (h) => {
    h.preventDefault(), h.button === 2 && t.editable && setTimeout(() => {
      if (t.panHelper.moved || t.ptState !== i.Idle && t.ptState !== i.Pan) return;
      const m = h.target;
      J(m) && !m.classList.contains("selected") && t.selectNode(m), t.bus.fire("showContextMenu", h);
    }, 200);
  }, x = (h) => {
    if (h.stopPropagation(), h.preventDefault(), h.ctrlKey || h.metaKey) return Ae(t, h);
    if (h.shiftKey) return t.move(-h.deltaY, 0);
    t.move(-h.deltaX, -h.deltaY);
  }, v = (h) => {
    h.code === "Space" && (t.spacePressed = !0, t.container.classList.add("space-pressed"));
  }, C = (h) => {
    h.code === "Space" && (t.spacePressed = !1, t.container.classList.remove("space-pressed"));
  };
  return vt([
    { dom: n, evt: "pointerdown", func: u },
    { dom: n, evt: "pointermove", func: g },
    { dom: n, evt: "pointerup", func: y },
    { dom: n, evt: "pointercancel", func: p },
    { dom: n, evt: "click", func: f },
    { dom: n, evt: "contextmenu", func: b },
    { dom: n, evt: "wheel", func: typeof t.handleWheel == "function" ? t.handleWheel : x },
    { dom: n, evt: "blur", func: p },
    { dom: n, evt: "keydown", func: v },
    { dom: n, evt: "keyup", func: C }
  ]);
}
function $e() {
  return {
    handlers: {},
    addListener: function(t, e) {
      this.handlers[t] === void 0 && (this.handlers[t] = []), this.handlers[t].push(e);
    },
    fire: function(t, ...e) {
      if (this.handlers[t] instanceof Array) {
        const n = this.handlers[t];
        for (let o = 0; o < n.length; o++)
          n[o](...e);
      }
    },
    removeListener: function(t, e) {
      if (!this.handlers[t]) return;
      const n = this.handlers[t];
      if (!e)
        n.length = 0;
      else if (n.length)
        for (let o = 0; o < n.length; o++)
          n[o] === e && this.handlers[t].splice(o, 1);
    }
  };
}
const k = "http://www.w3.org/2000/svg", U = function(t) {
  const e = t.clientWidth, n = t.clientHeight, o = t.dataset, s = Number(o.x), i = Number(o.y), r = o.anchor;
  let c = s;
  r === "middle" ? c = s - e / 2 : r === "end" && (c = s - e), t.style.left = `${c}px`, t.style.top = `${i - n / 2}px`, t.style.visibility = "visible";
}, Q = function(t, e, n, o) {
  const { anchor: s = "middle", color: i, dataType: r, svgId: c } = o, a = document.createElement("div");
  a.className = "svg-label", a.style.color = i || "#666";
  const d = "label-" + c;
  return a.id = d, a.innerHTML = t, a.dataset.type = r, a.dataset.svgId = c, a.dataset.x = e.toString(), a.dataset.y = n.toString(), a.dataset.anchor = s, a;
}, Dt = function(t, e, n) {
  const o = document.createElementNS(k, "path");
  return D(o, {
    d: t,
    stroke: e || "#666",
    fill: "none",
    "stroke-width": n
  }), o;
}, R = function(t) {
  const e = document.createElementNS(k, "svg");
  return e.setAttribute("class", t), e.setAttribute("overflow", "visible"), e;
}, gt = function() {
  const t = document.createElementNS(k, "line");
  return t.setAttribute("stroke", "#4dc4ff"), t.setAttribute("fill", "none"), t.setAttribute("stroke-width", "2"), t.setAttribute("opacity", "0.45"), t;
}, He = function(t, e, n, o) {
  const s = document.createElementNS(k, "g");
  return [
    {
      name: "line",
      d: t
    },
    {
      name: "arrow1",
      d: e
    },
    {
      name: "arrow2",
      d: n
    }
  ].forEach((r, c) => {
    const a = r.d, d = document.createElementNS(k, "path"), f = {
      d: a,
      stroke: o?.stroke || "rgb(227, 125, 116)",
      fill: "none",
      "stroke-linecap": o?.strokeLinecap || "cap",
      "stroke-width": String(o?.strokeWidth || "2")
    };
    o?.opacity !== void 0 && (f.opacity = String(o.opacity)), D(d, f), c === 0 && d.setAttribute("stroke-dasharray", o?.strokeDasharray || "8,2");
    const l = document.createElementNS(k, "path");
    D(l, {
      d: a,
      stroke: "transparent",
      fill: "none",
      "stroke-width": "15"
    }), s.appendChild(l), s.appendChild(d), s[r.name] = d;
  }), s;
}, Lt = function(t, e, n) {
  if (!e) return;
  const o = n.label;
  e.style.opacity = "0";
  const s = e.cloneNode(!0);
  t.nodes.appendChild(s), s.id = "input-box", s.textContent = o, s.contentEditable = "plaintext-only", s.spellcheck = !1, s.style.cssText = `
    left:${e.style.left};
    top:${e.style.top}; 
    max-width: 200px;
  `, Et(s), t.scrollIntoView(s), s.addEventListener("keydown", (i) => {
    if (i.stopPropagation(), i.isComposing) return;
    const r = i.key;
    if (r === "Enter" || r === "Tab") {
      if (i.shiftKey) return;
      i.preventDefault(), s.blur(), t.container.focus();
    }
  }), s.addEventListener("blur", () => {
    if (!s) return;
    const i = s.innerText?.trim() || "";
    i === "" ? n.label = o : n.label = i, e.style.opacity = "1", s.remove(), i !== o && (t.markdown ? e.innerHTML = t.markdown(n.label, n) : e.textContent = n.label, U(e), "parent" in n ? t.bus.fire("operation", {
      name: "finishEditSummary",
      obj: n
    }) : t.bus.fire("operation", {
      name: "finishEditArrowLabel",
      obj: n
    }));
  });
}, Ie = function(t) {
  const e = this.map.querySelector("me-root"), n = e.offsetTop, o = e.offsetLeft, s = e.offsetWidth, i = e.offsetHeight, r = this.map.querySelectorAll("me-main > me-wrapper");
  this.lines.innerHTML = "";
  for (let c = 0; c < r.length; c++) {
    const a = r[c], d = a.querySelector("me-tpc"), { offsetLeft: f, offsetTop: l } = P(this.nodes, d), u = d.offsetWidth, g = d.offsetHeight, y = a.parentNode.className, p = this.generateMainBranch({ pT: n, pL: o, pW: s, pH: i, cT: l, cL: f, cW: u, cH: g, direction: y, containerHeight: this.nodes.offsetHeight }), b = this.theme.palette, x = d.nodeObj.branchColor || b[c % b.length];
    if (d.style.borderColor = x, this.lines.appendChild(Dt(p, x, "3")), t && t !== a)
      continue;
    const v = R("subLines"), C = a.lastChild;
    C.tagName === "svg" && C.remove(), a.appendChild(v), Mt(this, v, x, a, y, !0);
  }
  this.labelContainer.innerHTML = "", this.renderArrow(), this.renderSummary(), this.bus.fire("linkDiv");
}, Mt = function(t, e, n, o, s, i) {
  const r = o.firstChild, c = o.children[1].children;
  if (c.length === 0) return;
  const a = r.offsetTop, d = r.offsetLeft, f = r.offsetWidth, l = r.offsetHeight;
  for (let u = 0; u < c.length; u++) {
    const g = c[u], y = g.firstChild, p = y.offsetTop, b = y.offsetLeft, x = y.offsetWidth, v = y.offsetHeight, C = y.firstChild.nodeObj.branchColor || n, T = t.generateSubBranch({ pT: a, pL: d, pW: f, pH: l, cT: p, cL: b, cW: x, cH: v, direction: s, isFirst: i });
    e.appendChild(Dt(T, C, "2"));
    const h = y.children[1];
    if (h) {
      if (!h.expanded) continue;
    } else
      continue;
    Mt(t, e, C, g, s);
  }
}, Oe = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169394918" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2021" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M851.91168 328.45312c-59.97056 0-108.6208 48.47104-108.91264 108.36992l-137.92768 38.4a109.14304 109.14304 0 0 0-63.46752-46.58688l1.39264-137.11872c47.29344-11.86816 82.31936-54.66624 82.31936-105.64096 0-60.15488-48.76288-108.91776-108.91776-108.91776s-108.91776 48.76288-108.91776 108.91776c0 49.18784 32.60928 90.75712 77.38368 104.27392l-1.41312 138.87488a109.19936 109.19936 0 0 0-63.50336 48.55808l-138.93632-39.48544 0.01024-0.72704c0-60.15488-48.76288-108.91776-108.91776-108.91776s-108.91776 48.75776-108.91776 108.91776c0 60.15488 48.76288 108.91264 108.91776 108.91264 39.3984 0 73.91232-20.92032 93.03552-52.2496l139.19232 39.552-0.00512 0.2304c0 25.8304 9.00096 49.5616 24.02816 68.23424l-90.14272 132.63872a108.7488 108.7488 0 0 0-34.2528-5.504c-60.15488 0-108.91776 48.768-108.91776 108.91776 0 60.16 48.76288 108.91776 108.91776 108.91776 60.16 0 108.92288-48.75776 108.92288-108.91776 0-27.14624-9.9328-51.968-26.36288-71.04l89.04704-131.03104a108.544 108.544 0 0 0 37.6832 6.70208 108.672 108.672 0 0 0 36.48512-6.272l93.13792 132.57216a108.48256 108.48256 0 0 0-24.69888 69.0688c0 60.16 48.768 108.92288 108.91776 108.92288 60.16 0 108.91776-48.76288 108.91776-108.92288 0-60.14976-48.75776-108.91776-108.91776-108.91776a108.80512 108.80512 0 0 0-36.69504 6.3488l-93.07136-132.48a108.48768 108.48768 0 0 0 24.79616-72.22784l136.09984-37.888c18.99008 31.93856 53.84192 53.3504 93.69088 53.3504 60.16 0 108.92288-48.75776 108.92288-108.91264-0.00512-60.15488-48.77312-108.92288-108.92288-108.92288z" p-id="2022"></path></svg>', Be = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169375313" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="1775" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M639 463.30000001L639 285.1c0-36.90000001-26.4-68.5-61.3-68.5l-150.2 0c-1.5 0-3 0.1-4.5 0.3-10.2-38.7-45.5-67.3-87.5-67.3-50 0-90.5 40.5-90.5 90.5s40.5 90.5 90.5 90.5c42 0 77.3-28.6 87.5-67.39999999 1.4 0.3 2.9 0.4 4.5 0.39999999L577.7 263.6c6.8 0 14.3 8.9 14.3 21.49999999l0 427.00000001c0 12.7-7.40000001 21.5-14.30000001 21.5l-150.19999999 0c-1.5 0-3 0.2-4.5 0.4-10.2-38.8-45.5-67.3-87.5-67.3-50 0-90.5 40.5-90.5 90.4 0 49.9 40.5 90.6 90.5 90.59999999 42 0 77.3-28.6 87.5-67.39999999 1.4 0.2 2.9 0.4 4.49999999 0.4L577.7 780.7c34.80000001 0 61.3-31.6 61.3-68.50000001L639 510.3l79.1 0c10.4 38.5 45.49999999 67 87.4 67 50 0 90.5-40.5 90.5-90.5s-40.5-90.5-90.5-90.5c-41.79999999 0-77.00000001 28.4-87.4 67L639 463.30000001z" fill="currentColor" p-id="1776"></path></svg>', Ye = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169667709" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="3037" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M385 560.69999999L385 738.9c0 36.90000001 26.4 68.5 61.3 68.5l150.2 0c1.5 0 3-0.1 4.5-0.3 10.2 38.7 45.5 67.3 87.5 67.3 50 0 90.5-40.5 90.5-90.5s-40.5-90.5-90.5-90.5c-42 0-77.3 28.6-87.5 67.39999999-1.4-0.3-2.9-0.4-4.5-0.39999999L446.3 760.4c-6.8 0-14.3-8.9-14.3-21.49999999l0-427.00000001c0-12.7 7.40000001-21.5 14.30000001-21.5l150.19999999 0c1.5 0 3-0.2 4.5-0.4 10.2 38.8 45.5 67.3 87.5 67.3 50 0 90.5-40.5 90.5-90.4 0-49.9-40.5-90.6-90.5-90.59999999-42 0-77.3 28.6-87.5 67.39999999-1.4-0.2-2.9-0.4-4.49999999-0.4L446.3 243.3c-34.80000001 0-61.3 31.6-61.3 68.50000001L385 513.7l-79.1 0c-10.4-38.5-45.49999999-67-87.4-67-50 0-90.5 40.5-90.5 90.5s40.5 90.5 90.5 90.5c41.79999999 0 77.00000001-28.4 87.4-67L385 560.69999999z" fill="currentColor" p-id="3038"></path></svg>', We = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169402629" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2170" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M639.328 416c8.032 0 16.096-3.008 22.304-9.056l202.624-197.184-0.8 143.808c-0.096 17.696 14.144 32.096 31.808 32.192 0.064 0 0.128 0 0.192 0 17.6 0 31.904-14.208 32-31.808l1.248-222.208c0-0.672-0.352-1.248-0.384-1.92 0.032-0.512 0.288-0.896 0.288-1.408 0.032-17.664-14.272-32-31.968-32.032L671.552 96l-0.032 0c-17.664 0-31.968 14.304-32 31.968C639.488 145.632 653.824 160 671.488 160l151.872 0.224-206.368 200.8c-12.672 12.32-12.928 32.608-0.64 45.248C622.656 412.736 630.976 416 639.328 416z" p-id="2171"></path><path d="M896.032 639.552 896.032 639.552c-17.696 0-32 14.304-32.032 31.968l-0.224 151.872-200.832-206.4c-12.32-12.64-32.576-12.96-45.248-0.64-12.672 12.352-12.928 32.608-0.64 45.248l197.184 202.624-143.808-0.8c-0.064 0-0.128 0-0.192 0-17.6 0-31.904 14.208-32 31.808-0.096 17.696 14.144 32.096 31.808 32.192l222.24 1.248c0.064 0 0.128 0 0.192 0 0.64 0 1.12-0.32 1.76-0.352 0.512 0.032 0.896 0.288 1.408 0.288l0.032 0c17.664 0 31.968-14.304 32-31.968L928 671.584C928.032 653.952 913.728 639.584 896.032 639.552z" p-id="2172"></path><path d="M209.76 159.744l143.808 0.8c0.064 0 0.128 0 0.192 0 17.6 0 31.904-14.208 32-31.808 0.096-17.696-14.144-32.096-31.808-32.192L131.68 95.328c-0.064 0-0.128 0-0.192 0-0.672 0-1.248 0.352-1.888 0.384-0.448 0-0.8-0.256-1.248-0.256 0 0-0.032 0-0.032 0-17.664 0-31.968 14.304-32 31.968L96 352.448c-0.032 17.664 14.272 32 31.968 32.032 0 0 0.032 0 0.032 0 17.664 0 31.968-14.304 32-31.968l0.224-151.936 200.832 206.4c6.272 6.464 14.624 9.696 22.944 9.696 8.032 0 16.096-3.008 22.304-9.056 12.672-12.32 12.96-32.608 0.64-45.248L209.76 159.744z" p-id="2173"></path><path d="M362.368 617.056l-202.624 197.184 0.8-143.808c0.096-17.696-14.144-32.096-31.808-32.192-0.064 0-0.128 0-0.192 0-17.6 0-31.904 14.208-32 31.808l-1.248 222.24c0 0.704 0.352 1.312 0.384 2.016 0 0.448-0.256 0.832-0.256 1.312-0.032 17.664 14.272 32 31.968 32.032L352.448 928c0 0 0.032 0 0.032 0 17.664 0 31.968-14.304 32-31.968s-14.272-32-31.968-32.032l-151.936-0.224 206.4-200.832c12.672-12.352 12.96-32.608 0.64-45.248S375.008 604.704 362.368 617.056z" p-id="2174"></path></svg>', Re = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169573443" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2883" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M514.133333 488.533333m-106.666666 0a106.666667 106.666667 0 1 0 213.333333 0 106.666667 106.666667 0 1 0-213.333333 0Z" fill="currentColor" p-id="2884"></path><path d="M512 64C264.533333 64 64 264.533333 64 512c0 236.8 183.466667 428.8 416 445.866667v-134.4c-53.333333-59.733333-200.533333-230.4-200.533333-334.933334 0-130.133333 104.533333-234.666667 234.666666-234.666666s234.666667 104.533333 234.666667 234.666666c0 61.866667-49.066667 153.6-145.066667 270.933334l-59.733333 68.266666V960C776.533333 942.933333 960 748.8 960 512c0-247.466667-200.533333-448-448-448z" fill="currentColor" p-id="2885"></path></svg>', Xe = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169419447" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2480" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M863.328 482.56l-317.344-1.12L545.984 162.816c0-17.664-14.336-32-32-32s-32 14.336-32 32l0 318.4L159.616 480.064c-0.032 0-0.064 0-0.096 0-17.632 0-31.936 14.24-32 31.904C127.424 529.632 141.728 544 159.392 544.064l322.592 1.152 0 319.168c0 17.696 14.336 32 32 32s32-14.304 32-32l0-318.944 317.088 1.12c0.064 0 0.096 0 0.128 0 17.632 0 31.936-14.24 32-31.904C895.264 496.992 880.96 482.624 863.328 482.56z" p-id="2481"></path></svg>', Ve = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169426515" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2730" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M863.744 544 163.424 544c-17.664 0-32-14.336-32-32s14.336-32 32-32l700.32 0c17.696 0 32 14.336 32 32S881.44 544 863.744 544z" p-id="2731"></path></svg>', ze = {
  side: Oe,
  left: Be,
  right: Ye,
  full: We,
  living: Re,
  zoomin: Xe,
  zoomout: Ve
}, I = (t, e) => {
  const n = document.createElement("span");
  return n.id = t, n.innerHTML = ze[e], n;
};
function Fe(t) {
  const e = document.createElement("div"), n = I("fullscreen", "full"), o = I("toCenter", "living"), s = I("zoomout", "zoomout"), i = I("zoomin", "zoomin");
  e.appendChild(n), e.appendChild(o), e.appendChild(s), e.appendChild(i), e.className = "mind-elixir-toolbar rb";
  let r = null;
  const c = () => {
    const d = t.container.getBoundingClientRect(), f = st(t.map.style.transform), l = d.width / 2, u = d.height / 2, g = (l - f.x) / t.scaleVal, y = (u - f.y) / t.scaleVal;
    r = {
      containerRect: d,
      currentTransform: f,
      mapCenterX: g,
      mapCenterY: y
    };
  }, a = () => {
    if (r) {
      const d = t.container.getBoundingClientRect(), f = d.width / 2, l = d.height / 2, u = f - r.mapCenterX * t.scaleVal, g = l - r.mapCenterY * t.scaleVal, y = u - r.currentTransform.x, p = g - r.currentTransform.y;
      t.move(y, p);
    }
  };
  return t.el.addEventListener("fullscreenchange", a), n.onclick = () => {
    c(), document.fullscreenElement !== t.el ? t.el.requestFullscreen() : document.exitFullscreen();
  }, o.onclick = () => {
    t.toCenter();
  }, s.onclick = () => {
    t.scale(t.scaleVal - t.scaleSensitivity);
  }, i.onclick = () => {
    t.scale(t.scaleVal + t.scaleSensitivity);
  }, e;
}
function Ge(t) {
  const e = document.createElement("div"), n = I("tbltl", "left"), o = I("tbltr", "right"), s = I("tblts", "side");
  return e.appendChild(n), e.appendChild(o), e.appendChild(s), e.className = "mind-elixir-toolbar lt", n.onclick = () => {
    t.initLeft();
  }, o.onclick = () => {
    t.initRight();
  }, s.onclick = () => {
    t.initSide();
  }, e;
}
function je(t) {
  t.container.append(Fe(t)), t.container.append(Ge(t));
}
function Pt({ pT: t, pL: e, pW: n, pH: o, cT: s, cL: i, cW: r, cH: c, direction: a, containerHeight: d }) {
  let f = e + n / 2;
  const l = t + o / 2;
  let u;
  a === B.LHS ? u = i + r : u = i;
  const g = s + c / 2, p = (1 - Math.abs(g - l) / d) * 0.25 * (n / 2);
  return a === B.LHS ? f = f - n / 10 - p : f = f + n / 10 + p, `M ${f} ${l} Q ${f} ${g} ${u} ${g}`;
}
function kt({ pT: t, pL: e, pW: n, pH: o, cT: s, cL: i, cW: r, cH: c, direction: a, isFirst: d }) {
  const f = parseInt(this.container.style.getPropertyValue("--node-gap-x"));
  let l = 0, u = 0;
  d ? l = t + o / 2 : l = t + o;
  const g = s + c;
  let y = 0, p = 0, b = 0;
  const x = Math.abs(l - g) / 300 * f;
  return a === B.LHS ? (b = e, y = b + f, p = b - f, u = i + f, `M ${y} ${l} C ${b} ${l} ${b + x} ${g} ${p} ${g} H ${u}`) : (b = e + n, y = b - f, p = b + f, u = i + r - f, `M ${y} ${l} C ${b} ${l} ${b - x} ${g} ${p} ${g} H ${u}`);
}
const _e = function(t, e = !0) {
  this.theme = t, this.generateMainBranch = this.theme.generateMainBranch || Pt, this.generateSubBranch = this.theme.generateSubBranch || kt;
  const o = {
    ...(this.theme.type === "dark" ? nt : et).cssVar,
    ...this.theme.cssVar
  };
  this.compact && (o["--node-gap-x"] = "15px", o["--node-gap-y"] = "2px", o["--main-gap-x"] = "30px", o["--main-gap-y"] = "6px");
  const s = Object.keys(o);
  for (let i = 0; i < s.length; i++) {
    const r = s[i];
    this.container.style.setProperty(r, o[r]);
  }
  e && this.refresh();
}, qe = function(t) {
  this.compact = t, this.theme && this.changeTheme(this.theme);
}, Ue = function(t) {
  return {
    dom: t,
    moved: !1,
    // differentiate click and move
    pointerdown: !1,
    lastX: 0,
    lastY: 0,
    handlePointerMove(e) {
      if (this.pointerdown) {
        this.moved = !0;
        const n = e.clientX - this.lastX, o = e.clientY - this.lastY;
        this.lastX = e.clientX, this.lastY = e.clientY, this.cb && this.cb(n, o);
      }
    },
    handlePointerDown(e) {
      e.button === 0 && (this.pointerdown = !0, this.lastX = e.clientX, this.lastY = e.clientY, this.dom.setPointerCapture(e.pointerId));
    },
    handleClear(e) {
      this.pointerdown = !1, e.pointerId !== void 0 && this.dom.releasePointerCapture(e.pointerId);
    },
    cb: null,
    init(e, n) {
      this.cb = n, this.handleClear = this.handleClear.bind(this), this.handlePointerMove = this.handlePointerMove.bind(this), this.handlePointerDown = this.handlePointerDown.bind(this), this.destroy = vt([
        { dom: e, evt: "pointermove", func: this.handlePointerMove },
        { dom: e, evt: "pointerleave", func: this.handleClear },
        { dom: e, evt: "pointerup", func: this.handleClear },
        { dom: this.dom, evt: "pointerdown", func: this.handlePointerDown }
      ]);
    },
    destroy: null,
    clear() {
      this.moved = !1, this.pointerdown = !1;
    }
  };
}, mt = {
  create: Ue
}, At = "#4dc4ff";
function Nt(t, e, n, o, s, i, r, c) {
  return {
    x: t / 8 + n * 3 / 8 + s * 3 / 8 + r / 8,
    y: e / 8 + o * 3 / 8 + i * 3 / 8 + c / 8
  };
}
function Ke(t, e, n) {
  t && (t.dataset.x = e.toString(), t.dataset.y = n.toString(), U(t));
}
function O(t, e, n, o, s) {
  D(t, {
    x1: e + "",
    y1: n + "",
    x2: o + "",
    y2: s + ""
  });
}
function tt(t, e, n, o, s, i, r, c, a, d) {
  const f = `M ${e} ${n} C ${o} ${s} ${i} ${r} ${c} ${a}`;
  t.line.setAttribute("d", f);
  const l = d.style || {};
  t.line.setAttribute("stroke", l.stroke || "rgb(227, 125, 116)"), t.line.setAttribute("stroke-width", String(l.strokeWidth || "2")), t.line.setAttribute("stroke-dasharray", l.strokeDasharray || "8,2"), t.line.setAttribute("stroke-linecap", l.strokeLinecap || "cap"), l.opacity !== void 0 && l.opacity !== null && l.opacity !== "" ? t.line.setAttribute("opacity", String(l.opacity)) : t.line.removeAttribute("opacity");
  const u = t.querySelectorAll('path[stroke="transparent"]');
  u.length > 0 && u[0].setAttribute("d", f);
  const g = _(i, r, c, a);
  if (g) {
    const x = `M ${g.x1} ${g.y1} L ${c} ${a} L ${g.x2} ${g.y2}`;
    t.arrow1.setAttribute("d", x), u.length > 1 && u[1].setAttribute("d", x), t.arrow1.setAttribute("stroke", l.stroke || "rgb(227, 125, 116)"), t.arrow1.setAttribute("stroke-width", String(l.strokeWidth || "2")), t.arrow1.setAttribute("stroke-linecap", l.strokeLinecap || "cap"), l.opacity !== void 0 && l.opacity !== null && l.opacity !== "" ? t.arrow1.setAttribute("opacity", String(l.opacity)) : t.arrow1.removeAttribute("opacity");
  }
  if (d.bidirectional) {
    const x = _(o, s, e, n);
    if (x) {
      const v = `M ${x.x1} ${x.y1} L ${e} ${n} L ${x.x2} ${x.y2}`;
      t.arrow2.setAttribute("d", v), u.length > 2 && u[2].setAttribute("d", v);
    }
  } else
    t.arrow2.setAttribute("d", ""), u.length > 2 && u[2].setAttribute("d", "");
  t.arrow2.setAttribute("stroke", l.stroke || "rgb(227, 125, 116)"), t.arrow2.setAttribute("stroke-width", String(l.strokeWidth || "2")), t.arrow2.setAttribute("stroke-linecap", l.strokeLinecap || "cap"), l.opacity !== void 0 && l.opacity !== null && l.opacity !== "" ? t.arrow2.setAttribute("opacity", String(l.opacity)) : t.arrow2.removeAttribute("opacity");
  const { x: y, y: p } = Nt(e, n, o, s, i, r, c, a);
  t.labelEl && Ke(t.labelEl, y, p);
  const b = t.labelEl;
  b && (b.style.color = l.labelColor || "rgb(235, 95, 82)"), on(t);
}
function Y(t, e, n) {
  const { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = e.offsetWidth, r = e.offsetHeight, c = o + i / 2, a = s + r / 2, d = c + n.x, f = a + n.y;
  return {
    w: i,
    h: r,
    cx: c,
    cy: a,
    ctrlX: d,
    ctrlY: f
  };
}
function H(t) {
  let e, n;
  const o = (t.cy - t.ctrlY) / (t.ctrlX - t.cx);
  return o > t.h / t.w || o < -t.h / t.w ? t.cy - t.ctrlY < 0 ? (e = t.cx - t.h / 2 / o, n = t.cy + t.h / 2) : (e = t.cx + t.h / 2 / o, n = t.cy - t.h / 2) : t.cx - t.ctrlX < 0 ? (e = t.cx + t.w / 2, n = t.cy - t.w * o / 2) : (e = t.cx - t.w / 2, n = t.cy + t.w * o / 2), {
    x: e,
    y: n
  };
}
const $t = function(t, e, n) {
  const o = P(t.nodes, e), s = P(t.nodes, n), i = o.offsetLeft + e.offsetWidth / 2, r = o.offsetTop + e.offsetHeight / 2, c = s.offsetLeft + n.offsetWidth / 2, a = s.offsetTop + n.offsetHeight / 2, d = c - i, f = a - r, l = Math.sqrt(d * d + f * f), u = Math.max(50, Math.min(200, l * 0.3)), g = Math.abs(d), y = Math.abs(f);
  let p, b;
  if (l < 150) {
    const v = e.closest("me-main").className === "lhs" ? -1 : 1;
    p = { x: 200 * v, y: 0 }, b = { x: 200 * v, y: 0 };
  } else if (g > y * 1.5) {
    const v = d > 0 ? e.offsetWidth / 2 : -e.offsetWidth / 2, C = d > 0 ? -n.offsetWidth / 2 : n.offsetWidth / 2;
    p = { x: v + (d > 0 ? u : -u), y: 0 }, b = { x: C + (d > 0 ? -u : u), y: 0 };
  } else if (y > g * 1.5) {
    const v = f > 0 ? e.offsetHeight / 2 : -e.offsetHeight / 2, C = f > 0 ? -n.offsetHeight / 2 : n.offsetHeight / 2;
    p = { x: 0, y: v + (f > 0 ? u : -u) }, b = { x: 0, y: C + (f > 0 ? -u : u) };
  } else {
    const v = Math.atan2(f, d), C = e.offsetWidth / 2 * Math.cos(v), T = e.offsetHeight / 2 * Math.sin(v), h = -(n.offsetWidth / 2) * Math.cos(v), m = -(n.offsetHeight / 2) * Math.sin(v), w = u * 0.7 * (d > 0 ? 1 : -1), E = u * 0.7 * (f > 0 ? 1 : -1);
    p = { x: C + w, y: T + E }, b = { x: h - w, y: m - E };
  }
  return {
    delta1: { x: Math.round(p.x), y: Math.round(p.y) },
    delta2: { x: Math.round(b.x), y: Math.round(b.y) }
  };
}, rt = function(t, e, n, o, s) {
  if (!e || !n)
    return;
  if (!o.delta1 || !o.delta2) {
    const S = $t(t, e, n);
    o.delta1 = S.delta1, o.delta2 = S.delta2;
  }
  const i = Y(t, e, o.delta1), r = Y(t, n, o.delta2), { x: c, y: a } = H(i), { ctrlX: d, ctrlY: f } = i, { ctrlX: l, ctrlY: u } = r, { x: g, y } = H(r), p = _(l, u, g, y);
  if (!p) return;
  const b = `M ${p.x1} ${p.y1} L ${g} ${y} L ${p.x2} ${p.y2}`;
  let x = "";
  if (o.bidirectional) {
    const S = _(d, f, c, a);
    if (!S) return;
    x = `M ${S.x1} ${S.y1} L ${c} ${a} L ${S.x2} ${S.y2}`;
  }
  const v = He(`M ${c} ${a} C ${d} ${f} ${l} ${u} ${g} ${y}`, b, x, o.style), { x: C, y: T } = Nt(c, a, d, f, l, u, g, y), h = o.style?.labelColor || "rgb(235, 95, 82)", m = "a-" + o.id;
  v.id = m;
  const w = t.markdown ? t.markdown(o.label, o) : o.label, E = Q(w, C, T, {
    anchor: "middle",
    color: h,
    dataType: "arrow",
    svgId: m
  });
  v.labelEl = E, v.arrowObj = o, v.dataset.linkid = o.id, t.labelContainer.appendChild(E), t.arrowSvg.appendChild(v), U(E), s || (t.arrows.push(o), t.currentArrow = v, It(t, o, i, r));
}, Je = function(t, e, n = {}) {
  const o = {
    id: X(),
    label: "Custom Link",
    from: t.nodeObj.id,
    to: e.nodeObj.id,
    ...n
  };
  rt(this, t, e, o), this.bus.fire("operation", {
    name: "createArrow",
    obj: o
  });
}, Ze = function(t) {
  K(this);
  const e = { ...t, id: X() };
  rt(this, this.findEle(e.from), this.findEle(e.to), e), this.bus.fire("operation", {
    name: "createArrow",
    obj: e
  });
}, Qe = function(t) {
  let e;
  if (t ? e = t : e = this.currentArrow, !e) return;
  K(this);
  const n = e.arrowObj.id;
  this.arrows = this.arrows.filter((o) => o.id !== n), e.labelEl?.remove(), e.remove(), this.bus.fire("operation", {
    name: "removeArrow",
    obj: {
      id: n
    }
  });
}, tn = function(t) {
  this.currentArrow = t;
  const e = t.arrowObj, n = this.findEle(e.from), o = this.findEle(e.to), s = Y(this, n, e.delta1), i = Y(this, o, e.delta2);
  this.editable ? It(this, e, s, i) : Ht(t, At), this.bus.fire("selectArrow", e);
}, en = function() {
  K(this), this.currentArrow = null, this.bus.fire("unselectArrow");
}, Z = function(t, e) {
  const n = document.createElementNS(k, "path");
  return D(n, {
    d: t,
    stroke: e,
    fill: "none",
    "stroke-width": "6",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }), n;
}, Ht = function(t, e) {
  const n = document.createElementNS(k, "g");
  n.setAttribute("class", "arrow-highlight"), n.setAttribute("opacity", "0.45");
  const o = Z(t.line.getAttribute("d"), e);
  n.appendChild(o);
  const s = Z(t.arrow1.getAttribute("d"), e);
  if (n.appendChild(s), t.arrow2.getAttribute("d")) {
    const i = Z(t.arrow2.getAttribute("d"), e);
    n.appendChild(i);
  }
  t.insertBefore(n, t.firstChild);
}, nn = function(t) {
  const e = t.querySelector(".arrow-highlight");
  e && e.remove();
}, on = function(t) {
  const e = t.querySelector(".arrow-highlight");
  if (!e) return;
  const n = e.querySelectorAll("path");
  n.length >= 1 && n[0].setAttribute("d", t.line.getAttribute("d")), n.length >= 2 && n[1].setAttribute("d", t.arrow1.getAttribute("d")), n.length >= 3 && t.arrow2.getAttribute("d") && n[2].setAttribute("d", t.arrow2.getAttribute("d"));
}, K = function(t) {
  t.helper1?.destroy(), t.helper2?.destroy(), t.linkController.style.display = "none", t.P2.style.display = "none", t.P3.style.display = "none", t.currentArrow && nn(t.currentArrow);
}, It = function(t, e, n, o) {
  const { linkController: s, P2: i, P3: r, line1: c, line2: a, nodes: d, map: f, currentArrow: l, bus: u } = t;
  if (!l) return;
  s.style.display = "initial", i.style.display = "initial", r.style.display = "initial", d.appendChild(s), d.appendChild(i), d.appendChild(r), Ht(l, At);
  let { x: g, y } = H(n), { ctrlX: p, ctrlY: b } = n, { ctrlX: x, ctrlY: v } = o, { x: C, y: T } = H(o);
  i.style.cssText = `top:${b}px;left:${p}px;`, r.style.cssText = `top:${v}px;left:${x}px;`, O(c, g, y, p, b), O(a, x, v, C, T), t.helper1 = mt.create(i), t.helper2 = mt.create(r), t.helper1.init(f, (h, m) => {
    p = p + h / t.scaleVal, b = b + m / t.scaleVal;
    const w = H({ ...n, ctrlX: p, ctrlY: b });
    g = w.x, y = w.y, i.style.top = b + "px", i.style.left = p + "px", tt(l, g, y, p, b, x, v, C, T, e), O(c, g, y, p, b), e.delta1.x = Math.round(p - n.cx), e.delta1.y = Math.round(b - n.cy), u.fire("updateArrowDelta", e);
  }), t.helper2.init(f, (h, m) => {
    x = x + h / t.scaleVal, v = v + m / t.scaleVal;
    const w = H({ ...o, ctrlX: x, ctrlY: v });
    C = w.x, T = w.y, r.style.top = v + "px", r.style.left = x + "px", tt(l, g, y, p, b, x, v, C, T, e), O(a, x, v, C, T), e.delta2.x = Math.round(x - o.cx), e.delta2.y = Math.round(v - o.cy), u.fire("updateArrowDelta", e);
  });
};
function sn() {
  this.arrowSvg.innerHTML = "", this.labelContainer.querySelectorAll('.svg-label[data-type="arrow"]').forEach((e) => e.remove());
  for (let e = 0; e < this.arrows.length; e++) {
    const n = this.arrows[e];
    try {
      rt(this, this.findEle(n.from), this.findEle(n.to), n, !0);
    } catch {
    }
  }
  this.nodes.appendChild(this.arrowSvg);
}
function rn(t) {
  K(this), t && t.labelEl && Lt(this, t.labelEl, t.arrowObj);
}
function cn() {
  this.arrows = this.arrows.filter((t) => j(t.from, this.nodeData) && j(t.to, this.nodeData));
}
const ln = function(t, e) {
  const n = Yt(t);
  n.style && e.style && (e.style = Object.assign({}, n.style, e.style)), Object.assign(t, e);
  const o = this.arrowSvg.querySelector(`g[data-linkid="${t.id}"]`);
  if (o) {
    if (e.label !== void 0 && o.labelEl) {
      const r = this.markdown ? this.markdown(t.label, t) : t.label;
      o.labelEl.innerHTML = r;
    }
    const s = this.findEle(t.from), i = this.findEle(t.to);
    if (s && i) {
      if (!t.delta1 || !t.delta2) {
        const b = $t(this, s, i);
        t.delta1 = t.delta1 || b.delta1, t.delta2 = t.delta2 || b.delta2;
      }
      const r = Y(this, s, t.delta1), c = Y(this, i, t.delta2), { x: a, y: d } = H(r), { ctrlX: f, ctrlY: l } = r, { ctrlX: u, ctrlY: g } = c, { x: y, y: p } = H(c);
      tt(o, a, d, f, l, u, g, y, p, t), this.currentArrow?.arrowObj?.id === t.id && (this.P2.style.cssText = `top:${l}px;left:${f}px;`, this.P3.style.cssText = `top:${g}px;left:${u}px;`, O(this.line1, a, d, f, l), O(this.line2, u, g, y, p));
    }
  }
  this.bus.fire("operation", {
    name: "reshapeArrow",
    obj: t,
    origin: n
  });
}, an = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createArrow: Je,
  createArrowFrom: Ze,
  editArrowLabel: rn,
  removeArrow: Qe,
  renderArrow: sn,
  reshapeArrow: ln,
  selectArrow: tn,
  tidyArrow: cn,
  unselectArrow: en
}, Symbol.toStringTag, { value: "Module" })), hn = function(t) {
  if (t.length === 0) throw new Error("No selected node.");
  if (t.length === 1) {
    const a = t[0].nodeObj, d = t[0].nodeObj.parent;
    if (!d) throw new Error("Can not select root node.");
    const f = d.children.findIndex((l) => a === l);
    return {
      parent: d.id,
      start: f,
      end: f
    };
  }
  let e = 0;
  const n = t.map((a) => {
    let d = a.nodeObj;
    const f = [];
    for (; d.parent; ) {
      const l = d.parent, g = l.children?.indexOf(d);
      d = l, f.unshift({ node: d, index: g });
    }
    return f.length > e && (e = f.length), f;
  });
  let o = 0;
  t: for (; o < e; o++) {
    const a = n[0][o]?.node;
    for (let d = 1; d < n.length; d++)
      if (n[d][o]?.node !== a)
        break t;
  }
  if (!o) throw new Error("Can not select root node.");
  const s = n.map((a) => a[o - 1].index).sort(), i = s[0] || 0, r = s[s.length - 1] || 0, c = n[0][o - 1].node;
  if (!c.parent) throw new Error("Please select nodes in the same main topic.");
  return {
    parent: c.id,
    start: i,
    end: r
  };
}, dn = function(t) {
  const e = document.createElementNS(k, "g");
  return e.setAttribute("id", t), e;
}, yt = function(t, e) {
  const n = document.createElementNS(k, "path");
  return D(n, {
    d: t,
    stroke: e || "#666",
    fill: "none",
    "stroke-linecap": "round",
    "stroke-width": "2"
  }), n;
}, fn = (t) => t.parentElement.parentElement, un = function(t, { parent: e, start: n }) {
  const o = t.findEle(e), s = o.nodeObj;
  let i;
  return s.parent ? i = o.closest("me-main").className : i = t.findEle(s.children[n].id).closest("me-main").className, i;
}, ct = function(t, e) {
  const { id: n, label: o, parent: s, start: i, end: r, style: c } = e, { nodes: a, theme: d, summarySvg: f } = t, u = t.findEle(s).nodeObj, g = un(t, e);
  let y = 1 / 0, p = 0, b = 0, x = 0;
  for (let W = i; W <= r; W++) {
    const lt = u.children?.[W];
    if (!lt)
      return t.removeSummary(n), null;
    const V = fn(t.findEle(lt.id)), { offsetLeft: z, offsetTop: at } = P(a, V), ht = i === r ? 10 : 20;
    W === i && (b = at + ht), W === r && (x = at + V.offsetHeight - ht), z < y && (y = z), V.offsetWidth + z > p && (p = V.offsetWidth + z);
  }
  let v, C;
  const T = u.parent ? 10 : 0, h = b + T, m = x + T, w = (h + m) / 2, E = c?.stroke || d.cssVar["--color"], S = c?.labelColor || d.cssVar["--color"], M = "s-" + n, A = t.markdown ? t.markdown(o, e) : o;
  g === B.LHS ? (v = yt(`M ${y + 10} ${h} c -5 0 -10 5 -10 10 L ${y} ${m - 10} c 0 5 5 10 10 10 M ${y} ${w} h -10`, E), C = Q(A, y - 20, w, { anchor: "end", color: S, dataType: "summary", svgId: M })) : (v = yt(`M ${p - 10} ${h} c 5 0 10 5 10 10 L ${p} ${m - 10} c 0 5 -5 10 -10 10 M ${p} ${w} h 10`, E), C = Q(A, p + 20, w, { anchor: "start", color: S, dataType: "summary", svgId: M }));
  const $ = dn(M);
  return $.appendChild(v), t.labelContainer.appendChild(C), U(C), $.summaryObj = e, $.labelEl = C, f.appendChild($), $;
}, pn = function(t = {}) {
  if (!this.currentNodes) return;
  const { currentNodes: e, summaries: n, bus: o } = this, { parent: s, start: i, end: r } = hn(e), c = { id: X(), parent: s, start: i, end: r, label: "summary", style: t.style }, a = ct(this, c);
  n.push(c), this.editSummary(a), o.fire("operation", {
    name: "createSummary",
    obj: c
  });
}, gn = function(t) {
  const e = X(), n = { ...t, id: e };
  ct(this, n), this.summaries.push(n), this.bus.fire("operation", {
    name: "createSummary",
    obj: n
  });
}, mn = function(t) {
  const e = this.summaries.findIndex((n) => n.id === t);
  e > -1 && (this.summaries.splice(e, 1), this.nodes.querySelector("#s-" + t)?.remove(), this.nodes.querySelector("#label-s-" + t)?.remove()), this.bus.fire("operation", {
    name: "removeSummary",
    obj: { id: t }
  });
}, yn = function(t) {
  const e = t.labelEl;
  e && e.classList.add("selected"), this.currentSummary = t, this.bus.fire("selectSummary", t.summaryObj);
}, bn = function() {
  this.currentSummary?.labelEl?.classList.remove("selected"), this.currentSummary = null, this.bus.fire("unselectSummary");
}, vn = function() {
  this.summarySvg.innerHTML = "", this.summaries.forEach((t) => {
    try {
      ct(this, t);
    } catch {
    }
  }), this.nodes.insertAdjacentElement("beforeend", this.summarySvg);
}, xn = function(t) {
  t && t.labelEl && Lt(this, t.labelEl, t.summaryObj);
}, wn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createSummary: pn,
  createSummaryFrom: gn,
  editSummary: xn,
  removeSummary: mn,
  renderSummary: vn,
  selectSummary: yn,
  unselectSummary: bn
}, Symbol.toStringTag, { value: "Module" })), L = "http://www.w3.org/2000/svg";
function Cn(t, e) {
  const n = document.createElementNS(L, "svg");
  return D(n, {
    version: "1.1",
    xmlns: L,
    height: t,
    width: e
  }), n;
}
function Sn(t, e) {
  return (parseInt(t) - parseInt(e)) / 2;
}
function En(t, e, n, o) {
  const s = document.createElementNS(L, "g");
  let i = "";
  return t.text ? i = t.text.textContent : i = t.childNodes[0].textContent, i.split(`
`).forEach((c, a) => {
    const d = document.createElementNS(L, "text");
    D(d, {
      x: n + parseInt(e.paddingLeft) + "",
      y: o + parseInt(e.paddingTop) + Sn(e.lineHeight, e.fontSize) * (a + 1) + parseFloat(e.fontSize) * (a + 1) + "",
      "text-anchor": "start",
      "font-family": e.fontFamily,
      "font-size": `${e.fontSize}`,
      "font-weight": `${e.fontWeight}`,
      fill: `${e.color}`
    }), d.innerHTML = c, s.appendChild(d);
  }), s;
}
function Tn(t, e, n, o) {
  let s = "";
  t.nodeObj?.dangerouslySetInnerHTML ? s = t.nodeObj.dangerouslySetInnerHTML : t.text ? s = t.text.textContent : s = t.childNodes[0].textContent;
  const i = document.createElementNS(L, "foreignObject");
  D(i, {
    x: n + parseInt(e.paddingLeft) + "",
    y: o + parseInt(e.paddingTop) + "",
    width: e.width,
    height: e.height
  });
  const r = document.createElement("div");
  return D(r, {
    xmlns: "http://www.w3.org/1999/xhtml",
    style: `font-family: ${e.fontFamily}; font-size: ${e.fontSize}; font-weight: ${e.fontWeight}; color: ${e.color}; white-space: pre-wrap;`
  }), r.innerHTML = s, i.appendChild(r), i;
}
function Dn(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "rect");
  return D(i, {
    x: o + "",
    y: s + "",
    rx: n.borderRadius,
    ry: n.borderRadius,
    width: n.width,
    height: n.height,
    fill: n.backgroundColor,
    stroke: n.borderColor,
    "stroke-width": n.borderWidth
  }), i;
}
function F(t, e, n = !1) {
  const o = getComputedStyle(e), { offsetLeft: s, offsetTop: i } = P(t.nodes, e), r = document.createElementNS(L, "rect");
  D(r, {
    x: s + "",
    y: i + "",
    rx: o.borderRadius,
    ry: o.borderRadius,
    width: o.width,
    height: o.height,
    fill: o.backgroundColor,
    stroke: o.borderColor,
    "stroke-width": o.borderWidth
  });
  const c = document.createElementNS(L, "g");
  c.appendChild(r);
  let a;
  return n ? a = Tn(e, o, s, i) : a = En(e, o, s, i), c.appendChild(a), c;
}
function Ln(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "a"), r = document.createElementNS(L, "text");
  return D(r, {
    x: o + "",
    y: s + parseInt(n.fontSize) + "",
    "text-anchor": "start",
    "font-family": n.fontFamily,
    "font-size": `${n.fontSize}`,
    "font-weight": `${n.fontWeight}`,
    fill: `${n.color}`
  }), r.innerHTML = e.textContent, i.appendChild(r), i.setAttribute("href", e.href), i;
}
function Mn(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "image");
  return D(i, {
    x: o + "",
    y: s + "",
    width: n.width + "",
    height: n.height + "",
    href: e.src
  }), i;
}
const G = 100, Pn = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">', kn = (t, e = !1) => {
  const n = t.nodes, o = n.offsetHeight + G * 2, s = n.offsetWidth + G * 2, i = Cn(o + "px", s + "px"), r = document.createElementNS(L, "svg"), c = document.createElementNS(L, "rect");
  D(c, {
    x: "0",
    y: "0",
    width: `${s}`,
    height: `${o}`,
    fill: t.theme.cssVar["--bgcolor"]
  }), i.appendChild(c), n.querySelectorAll(".subLines").forEach((l) => {
    const u = l.cloneNode(!0), { offsetLeft: g, offsetTop: y } = P(n, l.parentElement);
    u.setAttribute("x", `${g}`), u.setAttribute("y", `${y}`), r.appendChild(u);
  });
  const a = n.querySelector(".lines")?.cloneNode(!0);
  a && r.appendChild(a);
  const d = n.querySelector(".topiclinks")?.cloneNode(!0);
  d && r.appendChild(d);
  const f = n.querySelector(".summary")?.cloneNode(!0);
  return f && r.appendChild(f), n.querySelectorAll("me-tpc").forEach((l) => {
    l.nodeObj.dangerouslySetInnerHTML ? r.appendChild(F(t, l, !e)) : (r.appendChild(Dn(t, l)), r.appendChild(F(t, l.text, !e)));
  }), n.querySelectorAll(".tags > span").forEach((l) => {
    r.appendChild(F(t, l));
  }), n.querySelectorAll(".icons > span").forEach((l) => {
    r.appendChild(F(t, l));
  }), n.querySelectorAll(".hyper-link").forEach((l) => {
    r.appendChild(Ln(t, l));
  }), n.querySelectorAll("img").forEach((l) => {
    r.appendChild(Mn(t, l));
  }), D(r, {
    x: G + "",
    y: G + "",
    overflow: "visible"
  }), i.appendChild(r), i;
}, An = (t, e) => (e && t.insertAdjacentHTML("afterbegin", "<style>" + e + "</style>"), Pn + t.outerHTML);
function Nn(t) {
  return new Promise((e, n) => {
    const o = new FileReader();
    o.onload = (s) => {
      e(s.target.result);
    }, o.onerror = (s) => {
      n(s);
    }, o.readAsDataURL(t);
  });
}
const $n = function(t = !1, e) {
  const n = kn(this, t), o = An(n, e);
  return new Blob([o], { type: "image/svg+xml" });
}, Hn = async function(t = !1, e) {
  const n = this.exportSvg(t, e), o = await Nn(n);
  return new Promise((s, i) => {
    const r = new Image();
    r.setAttribute("crossOrigin", "anonymous"), r.onload = () => {
      const c = document.createElement("canvas");
      c.width = r.width, c.height = r.height, c.getContext("2d").drawImage(r, 0, 0), c.toBlob(s, "image/png", 1);
    }, r.src = o, r.onerror = i;
  });
}, In = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  exportPng: Hn,
  exportSvg: $n
}, Symbol.toStringTag, { value: "Module" })), On = {}, Bn = {
  getObjById: j,
  generateNewObj: Bt,
  layout: jt,
  linkDiv: Ie,
  editTopic: Qt,
  createWrapper: Ut,
  createParent: Kt,
  createChildren: Jt,
  createTopic: Zt,
  findEle: Ct,
  changeTheme: _e,
  changeCompact: qe,
  ...Te,
  ...On,
  ...an,
  ...wn,
  ...In,
  init(t) {
    if (t = JSON.parse(JSON.stringify(t)), !t || !t.nodeData) return new Error("MindElixir: `data` is required");
    t.direction !== void 0 && (this.direction = t.direction), t.compact !== void 0 && (this.compact = t.compact), this.changeTheme(t.theme || this.theme, !1), t.meta && (this.meta = t.meta), this.nodeData = t.nodeData, ot(this.nodeData), this.arrows = t.arrows || [], this.summaries = t.summaries || [], this.tidyArrow(), this.toolBar && je(this), this.layout(), this.linkDiv(), this.toCenter();
  },
  destroy() {
    this.disposable.forEach((t) => t()), this.el && (this.el.innerHTML = ""), this.el = void 0, this.nodeData = void 0, this.arrows = void 0, this.summaries = void 0, this.currentArrow = void 0, this.currentNodes = void 0, this.currentSummary = void 0, this.theme = void 0, this.direction = void 0, this.bus = void 0, this.container = void 0, this.map = void 0, this.lines = void 0, this.linkController = void 0, this.arrowSvg = void 0, this.P2 = void 0, this.P3 = void 0, this.line1 = void 0, this.line2 = void 0, this.nodes = void 0, this.selection?.destroy(), this.selection = void 0;
  },
  /**
   * @public
   * @param {boolean} enable
   */
  enableMobileMultiSelect(t) {
    this.mobileMultiSelect = t;
  }
}, Yn = "5.14.0";
function Wn(t) {
  return {
    x: 0,
    y: 0,
    moved: !1,
    // differentiate click and move
    mousedown: !1,
    handlePointerDown(e) {
      this.moved = !1;
      const n = e.target, o = t.mouseSelectionButton === 0 ? 2 : 0, s = t.spacePressed && e.button === 0 && e.pointerType === "mouse", i = !t.editable || e.button === o && e.pointerType === "mouse" || e.pointerType === "touch";
      !s && !i || (this.x = e.clientX, this.y = e.clientY, n.className !== "circle" && n.contentEditable !== "plaintext-only" && (this.mousedown = !0, n.setPointerCapture(e.pointerId)));
    },
    handlePointerMove(e) {
      if (!this.mousedown || e.target.contentEditable === "plaintext-only" && !t.spacePressed) return !1;
      const n = e.clientX - this.x, o = e.clientY - this.y;
      return this.x = e.clientX, this.y = e.clientY, this.moved = !0, t.move(n, o), !0;
    },
    handlePointerUp(e) {
      if (!this.mousedown) return;
      const n = e.target;
      n.hasPointerCapture && n.hasPointerCapture(e.pointerId) && n.releasePointerCapture(e.pointerId), this.mousedown = !1;
    },
    clear() {
      this.mousedown = !1, this.moved = !1;
    }
  };
}
function N({
  el: t,
  direction: e,
  editable: n,
  contextMenu: o,
  toolBar: s,
  keypress: i,
  mouseSelectionButton: r,
  selectionContainer: c,
  before: a,
  newTopicName: d,
  allowUndo: f,
  generateMainBranch: l,
  generateSubBranch: u,
  overflowHidden: g,
  compact: y,
  theme: p,
  alignment: b,
  scaleSensitivity: x,
  scaleMax: v,
  scaleMin: C,
  handleWheel: T,
  markdown: h,
  imageProxy: m,
  pasteHandler: w,
  mobileMultiSelect: E
}) {
  let S = null;
  const M = Object.prototype.toString.call(t);
  if (M === "[object HTMLDivElement]" ? S = t : M === "[object String]" && (S = document.querySelector(t)), !S) throw new Error("MindElixir: el is not a valid element");
  S.style.position = "relative", S.innerHTML = "", this.el = S, this.disposable = [], this.before = a || {}, this.newTopicName = d || "New Node", this.contextMenu = o ?? !0, this.toolBar = s ?? !0, this.keypress = i ?? !0, this.mouseSelectionButton = r ?? 0, this.direction = e ?? 1, this.editable = n ?? !0, this.allowUndo = f ?? !0, this.scaleSensitivity = x ?? 0.1, this.scaleMax = v ?? 1.4, this.scaleMin = C ?? 0.2, this.generateMainBranch = l || Pt, this.generateSubBranch = u || kt, this.overflowHidden = g ?? !1, this.compact = y ?? !1, this.alignment = b ?? "root", this.handleWheel = T ?? !0, this.markdown = h || void 0, this.imageProxy = m || void 0, this.currentNodes = [], this.currentArrow = null, this.scaleVal = 1, this.tempDirection = null, this.mobileMultiSelect = E ?? !1, this.panHelper = Wn(this), this.bus = $e(), this.container = document.createElement("div"), this.selectionContainer = c || this.container, this.container.className = "map-container";
  const A = window.matchMedia("(prefers-color-scheme: dark)");
  this.theme = p || (A.matches ? nt : et);
  const $ = document.createElement("div");
  $.className = "map-canvas", this.map = $, this.container.setAttribute("tabindex", "0"), this.container.appendChild(this.map), this.el.appendChild(this.container), this.nodes = document.createElement("me-nodes"), this.lines = R("lines"), this.summarySvg = R("summary"), this.linkController = R("linkcontroller"), this.P2 = document.createElement("div"), this.P3 = document.createElement("div"), this.P2.className = this.P3.className = "circle", this.P2.style.display = this.P3.style.display = "none", this.line1 = gt(), this.line2 = gt(), this.linkController.appendChild(this.line1), this.linkController.appendChild(this.line2), this.arrowSvg = R("topiclinks"), this.labelContainer = document.createElement("div"), this.labelContainer.className = "label-container", this.map.appendChild(this.nodes), this.overflowHidden ? this.container.style.overflow = "hidden" : this.disposable.push(Ne(this)), w && (this.pasteHandler = w);
}
N.prototype = Bn;
Object.defineProperty(N.prototype, "currentNode", {
  get() {
    return this.currentNodes[this.currentNodes.length - 1];
  },
  enumerable: !0
});
N.LEFT = 0;
N.RIGHT = 1;
N.SIDE = 2;
N.THEME = et;
N.DARK_THEME = nt;
N.version = Yn;
N.E = Ct;
export {
  nt as DARK_THEME,
  Rn as LEFT,
  Xn as RIGHT,
  Vn as SIDE,
  et as THEME,
  N as default
};
